plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.tsx

```diff
@@ -0,0 +1,166 @@
+      },
+): string {
+  // Simplify the diagram output by hiding only the default namespace
+  return serializeEntityRef(ref)
+    .toString()
+    .toLowerCase()
+    .replace(':default/', ':');
+}
+
+/**
+ * Dynamically generates a diagram of a system, its assigned entities,
+ * and relationships of those entities.
+ */
+export function SystemDiagramCard() {
+  const { entity } = useEntity();
+  const currentSystemName = entity.metadata.name;
+  const currentSystemNode = simplifiedEntityName(entity);
+  const systemNodes = new Array<{ id: string }>();
+  const systemEdges = new Array<{ from: string; to: string; label: string }>();
+
+  const catalogApi = useApi(catalogApiRef);
+  const { loading, error, value: catalogResponse } = useAsync(() => {
+    return catalogApi.getEntities({
+      filter: {
+        kind: ['Component', 'API', 'Resource', 'System', 'Domain'],
+        ['spec.system']: currentSystemName,
+      },
+    });
+  }, [catalogApi]);
+
+  // pick out the system itself
+  systemNodes.push({
+    id: currentSystemNode,
+  });
+
+  // check if the system has an assigned domain
+  // even if the domain object doesn't exist in the catalog, display it in the map
+  const catalogItemDomain = getEntityRelations(entity, RELATION_PART_OF, {
+    kind: 'domain',
+  });
+  catalogItemDomain.forEach(foundDomain =>
+    systemNodes.push({
+      id: simplifiedEntityName(foundDomain),
+    }),
+  );
+  catalogItemDomain.forEach(foundDomain =>
+    systemEdges.push({
+      from: currentSystemNode,
+      to: simplifiedEntityName(foundDomain),
+      label: 'part of',
+    }),
+  );
+
+  if (catalogResponse && catalogResponse.items) {
+    for (const catalogItem of catalogResponse.items) {
+      systemNodes.push({
+        id: simplifiedEntityName(catalogItem),
+      });
+
+      // Check relations of the entity assigned to this system to see
+      // if it relates to other entities.
+      // Note those relations may, or may not, be explicitly
+      // assigned to the system.
+      const catalogItemRelations_partOf = getEntityRelations(
+        catalogItem,
+        RELATION_PART_OF,
+      );
+      catalogItemRelations_partOf.forEach(foundRelation =>
+        systemEdges.push({
+          from: simplifiedEntityName(catalogItem),
+          to: simplifiedEntityName(foundRelation),
+          label: 'part of',
+        }),
+      );
+
+      const catalogItemRelations_providesApi = getEntityRelations(
+        catalogItem,
+        RELATION_PROVIDES_API,
+      );
+
+      catalogItemRelations_providesApi.forEach(foundRelation =>
+        systemEdges.push({
+          from: simplifiedEntityName(catalogItem),
+          to: simplifiedEntityName(foundRelation),
+          label: 'provides API',
+        }),
+      );
+    }
+  }
+
+  if (loading) {
+    return <Progress />;
+  } else if (error) {
+    return (
+      <WarningPanel severity="error" title="Could not load system diagram">
+        {error.message}
+      </WarningPanel>
```

Comment:
This would be a good place to use `ResponseErrorPanel`, since the catalog emits those rich errors

Reply:
Hmm.  I'll admit to not being familiar with that component.  The code currently emits the rich errors via the screen cap from up earlier when I added warning panel:

![image](https://user-images.githubusercontent.com/33203301/112639321-8105d580-8e16-11eb-9978-7d3f55bdf5fc.png)

But happy to change this now to use `ResponseErrorPanel` instead.

Another one to maybe link back to the Error RFC https://github.com/backstage/backstage/issues/3781 and get into a Storybook example to show all the various different permutations of error styling, understand the nuance of each, and recommendations on when to use them. (which then might also help us refactor them more effectively!)

Reply:
Wait, that looks just like `ResponseErrorPanel` :) And it's rendered not by your code, but by the surrounding wrapper component.

Reply:
yeah - i have a feeling the warning panel calls it somehow. anyway, moved this code back to the raw ResponseErrorPanel per your note.